src/libraries/System.Private.CoreLib/src/System/IO/StringWriter.cs

```diff
@@ -151,22 +151,133 @@ public override void Write(string? value)
 
         public override void Write(StringBuilder? value)
         {
-            if (GetType() != typeof(StringWriter))
+            // This overload was added after the Write(char[], ...) overload, and so in case
+            // a derived type may have overridden it, we need to delegate to it, which the base does.
+            if (UseBase())
             {
-                // This overload was added after the Write(char[], ...) overload, and so in case
-                // a derived type may have overridden it, we need to delegate to it, which the base does.
                 base.Write(value);
-                return;
             }
+            else
+            {
+                _sb.Append(value);
+            }
+        }
+
+        #region Write(format) + WriteLine(format)
 
+        [System.Runtime.CompilerServices.MethodImpl(Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
+        private bool UseBase()
+        {
```

Comment:
This is adding additional checks where they weren't before.  Previously if this wasn't StringWriter, we'd just call base.Write, and the method it delegated to would end up handling the disposed check.  With your changes, it's now doing the disposed check here as well, unnecessarily.

I also don't think this UseBase method helps.  From my perspective, it hinders the clarity of what the caller is doing and why.  Please revert it and just use `GetType() != typeof(StringWriter)` in the call sites.  We can also replace these `if (!_isOpen) throw new ObjectDisposedException(...)` checks with `ObjectDisposedException.ThrowIf(!_isOpen, this);`.

Reply:
We'll likely also need to do argument validation before the disposed check rather than leaving it up to StringBuilder.